src/Function.cpp

```diff
@@ -497,6 +525,31 @@ Function::make_first(void)
 
 	if(CGOptions::func_attr_aligned() && rnd_flipcoin(FuncAttrAligned))
 		f->func_attr_aligned = true;
+
+	if(CGOptions::func_attr_flag() && rnd_flipcoin(FuncAttr)){
+		f->func_attr_flag = true;
+
+		vector<string>::iterator itr;
+		for(itr = f->attributes.begin(); itr < f->attributes.end(); itr++){
+			if(rnd_flipcoin(50))
+				f->func_attr.push_back(*itr);
+		}
+
+		if(rnd_flipcoin(50))
+			f->func_attr.push_back("used");
+		else
+			f->func_attr.push_back("unused");
+
+		if(rnd_flipcoin(20))
+			f->func_attr.push_back("visibility(\"default\")");
+		else if(rnd_flipcoin(30))
+			f->func_attr.push_back("visibility(\"hidden\")");
+		else if(rnd_flipcoin(40))
+			f->func_attr.push_back("visibility(\"protected\")");
+		else
+			f->func_attr.push_back("visibility(\"internal\")");
```

Comment:
Hm, apparently we have at least 2 attributes that are not of a simple boolean type. I'm suggesting to make step back a prepare a nicer infrastructure:

I'm suggesting to coming up with a `FunctionAttribute` class where the class will be a generator for each individual attribute. The class will be able to generate an attribute declaration and can have super classes for more complicated attributes like `aligned`, `visibility`, `no_sanitize`, `used/unused` and probably another that we'll see in the future.
Can you please design and use such a class?

Reply:
> Hm, apparently we have at least 2 attributes that are not of a simple boolean type. I'm suggesting to make step back a prepare a nicer infrastructure:
> 
> I'm suggesting to coming up with a `FunctionAttribute` class where the class will be a generator for each individual attribute. The class will be able to generate an attribute declaration and can have super classes for more complicated attributes like `aligned`, `visibility`, `no_sanitize`, `used/unused` and probably another that we'll see in the future.
> Can you please design and use such a class?

Can there be separate pull request for complicated attributes like `aligned`, `visibility`, `no_sanitize`, `used/unused`? This approach needs less time to get work done.

Remaining attributes are just setting bool and output. Is my approach from choosing attributes except `visibility` and `used/unused` from set of attributes stored in a vector not correct. Further, I planned to push function attributes like `alias`, `section`, `pure` and `const`. In this way, I planned to get deliverables mentioned in the timeline.

If not, I can try designing new class as you mentioned above.

Reply:
> If not, I can try designing new class as you mentioned above.

I would start with that. It will not be difficult and it will be more saner.

Reply:
Should I create class `FunctionAttribute` in Function class itself? or Separate files for `FunctionAttribute`

Reply:
Select what you think will be better.

Reply:
I created `FunctionAttribute` class in Function file itself. During function generation, i.e. `make_random`, attributes are generated using `generate_func_attr()` of class `FunctionAttribute` class and are stored which can be output during output phase.

Is it the correct way?

Reply:
No that's not what we want. We want to have a class instance for each individual attribute (and a probability will be passed). Then you will foreach all these instances and let you generate an attribute. Then you will need something like `class VisibilityFunctionAttribute: public FunctionAttribute` where you will override the function that generates to output (via `<<`).

Reply:
As per my understanding, `FunctionAttribute` will be base class with `virtual` methods. So that any derived class such as `VisibilityFunctionAttribute`, `Aligned`, `OutputAttribute`, `used/unused` can have their own definition of `virtual` methods declared in base class.

In future, if I want to add another type of function attribute, I will inherit base class and write my own definition. Is this what we want?

Reply:
I inherited `FunctionAttribute` class as `VisibilityFunctionAttribute` subclass which will generate `visibility` attribute during `make_random` of function and store in `output_func_attr` of respective function. In this way, during generation phase, desired attributes are created and in output phase, these attributes will output. Am I heading in correct path?

Reply:
Well, I would like to see a base class that will have a `probability` and a `name` variables.
The base will have virtual methods like `output_attr` (pure virtual). Then you'll inherit a `BooleanFunctionAttribute` that will override the method. I would like to see having `Function` having a static vector of instances of all these attribute. Then you will just `foraech` and attributes will be generated.

Reply:
> Well, I would like to see a base class that will have a `probability` and a `name` variables.
> The base will have virtual methods like `output_attr` (pure virtual). Then you'll inherit a `BooleanFunctionAttribute` that will override the method. I would like to see having `Function` having a static vector of instances of all these attribute. Then you will just `foraech` and attributes will be generated.

I have created a vector `output_func_attr` which contains all the instances of attributes generated.
In base class, I have declared a variable FuncAttrProb, a function named `generate_func_attr()` which is pure virtual and a vector ``common_func_attr` which will be used by base classes as per their requirement. 

In base classes, `generate_func_attr` selects attributes from `common_func_attr` and pushed into the vector of respective function.

I didn't get the significance of name variable

Please check latest squashed commit and suggest changes. I have created two base classes `BooleanFunctionAttribute` and `VisibilityFunctionAttribute`, when called generates respective attributes and add it to function's vector.

Reply:
You don't have to write long explanation, it's easier to express that via patch which I can easily review. So next time just write brief description. Thanks.

Reply:
Can you suggest how to use foreach on updated squashed commit? There is `for` loop for `BooleanFunctionAttribute`.

And also, due to squashed commit, I am unable to see previous comments done by you.